### ISSUE TITLE
Search service: label positioning for ch.swisstopo-vd.ortschaftenverzeichnis_plz

### ISSUE THREAD

**Issue by p1d1d1** [geoadmin/mf-chsdi3#2317]
From helpdesk:

 Bonjour,
J'ai été surpris de découvrir que vous mentionnez Couvet comme étant situé dans le canton de Vaud (voir capture d'écran en annexe) :-) Cela arrive quand on cherche par numéro postal (2108).
Meilleures salutations,

So, use case:
* open map.geo.admin.ch
* search for 2108 --> the first entry is "ZIP 2108 - Couver (VD)"

![image](https://cloud.githubusercontent.com/assets/1533649/20056599/63a7b35e-a4e7-11e6-8558-994accef4a9a.png)

The PIN is positioned in VD:

![image](https://cloud.githubusercontent.com/assets/1533649/20056648/a0c8e884-a4e7-11e6-9f12-b9cc749b1570.png)

The issue is that "Couvet" as locality is in NE. Now the positioning of the label in ch.swisstopo-vd.ortschaftenverzeichnis_plz is not really an issue, since it is in the polygon of ZIP 2108.

What is strange is the text in the search result (Couvet (VD)).

One could:

* delete the canton abbreviation beside the ZIP in the search result
* or use the points available in the original data for the positioning of the labels.

**Comment by p1d1d1** [comment]
@gjn any feedback on this?

**Comment by loicgasser** [comment]
>>or use the points available in the original data for the positioning of the labels.

``vd.gabmo_plz``` only has one geometry and those are always Polygons.

**Comment by gjn** [comment]
Couvet is a special case that the PLZ 2108 is in 2 cantons: VD and NE. 

The point is ST_PointOnSurface of the polygon - I assume it's some kind of Schwerpunkt. No, this point intersects with canton VD, not NE. That's why there's a VD in the text.
https://github.com/geoadmin/service-sphinxsearch/blob/master/conf/search.conf.part#L238

To account for multi-canton PLZ (have VD and NE in the label, which would be the most correct) would be quite complicated.

**Comment by p1d1d1** [comment]
The position of the label is on the original data, at least here: http://www.cadastre.ch/internet/kataster/de/home/services/service/plz.html

And it's not so nice to say that Couvet is in VD.

Please reconsider it.

**Comment by p1d1d1** [comment]
PS: @gjn the ST_PointOnSurface is calculated by you, I guess with some PostGIS functions.
IMHO you should use the position defined in the original data.

**Comment by ltclm** [comment]
agree with @p1d1d1 , this information should come with the original data, but it is not.

We are intersecting the plz geometry (polygon input replaced with "pointonsurface") with the canton geometry. For performance reasons this has been done with a single point geometry (pointonsurface) [1].
We could extent this query and do an polygon/polygon intersection, which will return multiple results in some cases. the result will be ordered descending by intersection area [2]. We could also show only the canton with the biggest intersection. couvet is covering almost the same surface in both cantons.

[1]
```sql
SELECT ak from tlm.swissboundaries_kantone k where st_intersects(ST_PointOnSurface(p.the_geom),k.the_geom)
// -> VD
```

[2]
```sql
select array_to_string(array_agg(ak order by st_area(st_intersection(p.the_geom,k.the_geom))desc),',') FROM vd.gabmo_plz p, tlm.swissboundaries_kantone k WHERE  st_intersects(p.the_geom,k.the_geom) 
// -> NE,VD
```

**Comment by ltclm** [comment]
[TestUrl](https://mf-geoadmin3.dev.bgdi.ch/?topic=ech&lang=de&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.swisstopo.zeitreihen,ch.bfs.gebaeude_wohnungs_register,ch.bav.haltestellen-oev,ch.swisstopo.swisstlm3d-wanderwege&layers_visibility=false,false,false,false&layers_timestamp=18641231,,,&sdf&X=190887.11&Y=556057.74&zoom=3)

**Comment by p1d1d1** [comment]
@ltclm first many thanks for your effort. Sorry but I still have some doubts.

In Chrome the result is quite weird:
![image](https://cloud.githubusercontent.com/assets/1533649/20170196/a42c21c6-a72b-11e6-95e0-3daebaec9a1f.png)

In IE seems ok:
![image](https://cloud.githubusercontent.com/assets/1533649/20170215/b55dfece-a72b-11e6-8191-c90fecc5b8c5.png)

But I still thinks that the position of the label is wrong. In the VD Data is:
![image](https://cloud.githubusercontent.com/assets/1533649/20170242/cdc4a328-a72b-11e6-8a0b-fcba82d511c0.png)

Now I don't know where you take the data from, but in all file available here: http://www.cadastre.ch/internet/kataster/de/home/services/service/plz.html
there is a table PLZO_OSNAMEPOS with X,Y of the point for the positioning of the label.

I really think we have to use the "original" data.

**Comment by ltclm** [comment]
This is a cache issue, try to reload the application in chrome.
our data comes from vd and there is no such information. 
@mariokeusen  can you ask the dataowner if they can add a label position ``PLZO_OSNAMEPOS``?

**Comment by gjn** [comment]
We miss this info the GDWH. So not possible.

But I think Marcel's approach is really good. It tells people that this PLZ is in 2 cantons - which is correct.

**Comment by p1d1d1** [comment]
PLZ  is in 2 cantons--> right, but the label position is wrong.

@gjn I did an extract of the dataset from the GDWH (link via business email). 

- In Interlis you have a table  OrtschaftsName_Pos

- in shapefiles you have OrtschaftsName_Pos

- in CSV you also have Couvet;2108;0;Val-de-Travers;NE;2538608;1197417

So the info is in GDWH

**Comment by p1d1d1** [comment]
@gjn @ltclm any feedback here?

**Comment by ltclm** [comment]
@mariokeusen will inform the dataowner

**Comment by p1d1d1** [comment]
For what?

**Comment by ltclm** [comment]
we have to change the source of our automatic data import. As soon as we get the new attributes, we will have to update the import scripts and use the exact label position for the sphinx query.

**Comment by AFoletti** [comment]
@ltclm I rapidly checked the data we use for the import and the needed label points are there indeed.
What we missed is the fact that the point geometries for label placement are NOT in the same table as the labels themselves.
In the PLZ workbench we should add the PLZO_OSNAMEPOS table and merge it with the PLZO_OSNAME. The rest of the process can stay exactly the same, but we will maybe need to adapt the target table.

**Comment by ltclm** [comment]
great thank you @AFoletti 
i will update the fme workbench, the table and the sphinx query.

**Comment by ltclm** [comment]
the sphinx index has now the exact label positions from plz dataset.
[TestUrl](https://mf-geoadmin3.dev.bgdi.ch/?swisssearch=2108)

**Comment by p1d1d1** [comment]
Perfect, many thanks.

**Comment by gjn** [comment]
You are great!

What needs to be done to deploy this? Please mention it in the deploy db.

**Comment by ltclm** [comment]
once this pr is merged, we have to deploy the new sphinx config to int and prod, and afterwards stopo_master.vd.gabmo_plz.
i have remarked it in the remember deploy document for 2016-11-30.

**Comment by p1d1d1** [comment]
any news here?

**Comment by p1d1d1** [comment]
@gjn this is still not fixed

**Comment by ltclm** [comment]
sorry this one got forgotten. 
there is a pr ready for the tomorrows deploy:
https://github.com/geoadmin/service-sphinxsearch/pull/325

**Comment by loicgasser** [comment]
Fixed

**Comment by p1d1d1** [comment]
This got fixed in search but label is still wrongly positioned on the map:

https://mf-geoadmin3.dev.bgdi.ch/?ch.swisstopo-vd.ortschaftenverzeichnis_plz=1F88B46B-18CE-42A5-96D5-DFE0005F2C2D&lang=it&topic=ech

**Comment by ltclm** [comment]
@p1d1d1 does it concern the label position in the wms service?

**Comment by p1d1d1** [comment]
I would say. Since this data has coordinates (actually the one used now in search service) for label positioning I think it makes sense to use these as anchor point for labels on the map.

**Comment by ltclm** [comment]
that's not possible with mapserver.
http://mapserver.org/mapfile/label.html

the label position can be configured as follows, the label point is calculated by mapserver label engine on the fly:
> POSITION [ul|uc|ur|cl|cc|cr|ll|lc|lr|auto]
Position of the label relative to the labeling point (layers only). First letter is “Y” position, second letter is “X” position. “Auto” tells MapServer to calculate a label position that will not interfere with other labels. With points, MapServer selects from the 8 outer positions (i.e. excluding cc). With polygons, MapServer selects from cc (added in MapServer 5.4), uc, lc, cl and cr as possible positions. With lines, it only uses lc or uc, until it finds a position that doesn’t collide with labels that have already been drawn. If all positions cause a conflict, then the label is not drawn (Unless the label’s FORCE a parameter is set to “true”). “Auto” placement is only available with cached labels.

**Comment by p1d1d1** [comment]
But if you have a point layer you can have the label close to the point, can't you?
Then this would need a group layer, one for the polygons and a point layer for labels, IMHO.

**Comment by ltclm** [comment]
yes, we can always add hacks, the question is : did somebody complain, is this a change request from the data owner or is it a nice to have enhancement discovered by accident?

**Comment by davidoesch** [comment]
Hi folks

The proper solution for POLYGON is a highlight of the polygon...not labeling the center: 
see https://github.com/geoadmin/mf-geoadmin3/issues/3873 

therefore: no need here to go further

**Comment by p1d1d1** [comment]
Highlight a polygon? I guess you didn't get the issue and it's not related to #3873.

**Comment by davidoesch** [comment]
user searches for ZIP: the result should *not* be a label but a perimeter... this is the issue.

**Comment by p1d1d1** [comment]
Not speaking of that, which is ok.
Open the plz layer. You'll notice there is a label in each polygon. These labels are positioned automatically by mapserver and not according what is defined in the source data.

**Comment by AFoletti** [comment]
What @p1d1d1 wants to change is simply where the labels for [ch.swisstopo-vd.ortschaftenverzeichnis_plz](https://s.geo.admin.ch/744802e8d1) are placed, in order to reflect the swisssearch (and datasource) positioning.
It looks to me like a quite not too complicated change in [this mapfile](https://github.com/geoadmin/wms-mapfile_include/blob/7881b0ba48635f85a9770765e8afdef3040e5c2f/ch.swisstopo-vd.ortschaftenverzeichnis_plz.map). Make it a group, use label placement for the second layer (point) and do nothing on said layer apart from showing a label.
I could very well be wrong, but to me it's 1 hour work including tests (layer is not even WMTS).

I will make the layer symbolization consistent with the underlying data. But then, yes, will have another WMS group instead of a simple layer

**Comment by davidoesch** [comment]
Now I get it.+1
